src/app/OperationalDeviceProxy.cpp

```diff
@@ -194,35 +233,37 @@ void OperationalDeviceProxy::EnqueueConnectionCallbacks(Callback::Callback<OnDev
     }
 }
 
-void OperationalDeviceProxy::DequeueConnectionSuccessCallbacks(bool executeCallback)
+void OperationalDeviceProxy::DequeueConnectionCallbacks(CHIP_ERROR error)
 {
     Cancelable ready;
-    mConnectionSuccess.DequeueAll(ready);
+    mConnectionFailure.DequeueAll(ready);
```

Comment:
This part is tricky.

Once we start calling callbacks, `this` might be dead because any callback can kill is.

The old code handled this by always calling the "don't call callbacks" dequeue function first and then calling the one that does call callbacks.  So on success we would do:
```
        device->DequeueConnectionFailureCallbacks(CHIP_NO_ERROR, /* executeCallback */ false);
        device->DequeueConnectionSuccessCallbacks(/* executeCallback */ true);
```
and on failure we would do:
```
    device->DequeueConnectionSuccessCallbacks(/* executeCallback */ false);
    device->DequeueConnectionFailureCallbacks(error, /* executeCallback */ true);
```
or so.

We should either keep the two separate functions for calling success/failure callbacks and sequence the calls to them correctly in this function or we should have two on-stack lists here, one for success and one for failure, DequeueAll into both up front, then start doing the loops that might call things.  And clearly document the "don't touch `this` after here" point in the function.

Reply:
Went with the on-stack lists approach.